quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/cli/Picocli.java

```diff
@@ -269,9 +274,65 @@ private static boolean hasConfigChanges() {
             }
         }
 
+        //check for defined quarkus raw build properties for UserStorageProvider extensions
+        if (QuarkusPropertiesConfigSource.getConfigurationFile() != null) {
```

Comment:
IMO, we should not worry about dealing with this situation at this momment.

As we discussed, ideally we should be able to query Quarkus if a property is a build time property. 

Including here a whitelist of build properties deserves a bit more discussion because. For instance, the possibility to add data sources is a capability we should support in order to deploy custom user providers. In this case, we should think about wrapping the data source config.

Reply:
I have to disagree. I still think the behaviour we have here with the whitelist is better for testing purposes _for now_, as long as we don't have first-class configuration options. 

First-class options would open up the "dependant options" space again, so not trivial. "Only allow/show options when the named additional datasource is there" is something we can't do as of now. A possible starting point to look at: we could have e.g. a build option `db-external-names=foo,bar,...` and only show and map the right way options like e.g. `kc-db-foo=mysql`, `kc-db-bar=postgres" and so on... so this is not trivial to achieve imo. at least not if we want to support the CLI with these and not have a ux issue on that matter. And for me definitely nothing for this ticket.

So, without first class support, I consider a dev who runs start --auto-build often to try things out before creating an optimized image for prod (and this is done often it seems from various other issues). 

The devex while doing this without the whitelist would be "start it, run the build bc. it detects quarkus.properties everytime even if only the username / url changed for the external datasource" and thus not good IMO. So at least until we support first class config for additional datasources, i think the whitelist approach is **good for now** and I am happy to change it when we either 

a) get sth from quarkus to check more fine-grained for build/runtime, or 
b) have first class support for additional datasources config on our side. 

Both approaches are some time down the road i guess, and it's not that hard to change the impl later on, so as a stop-gap solution I'd prefer keeping it that way.

Reply:
would be nice to get a third opinion, as ours seem to not go together here ;) cc @vmuzikar @andreaTP @stianst

Reply:
I'm not hard on that if others are OK. Looks like the extension story keeps knocking on the door :)

If we can avoid temporary code, better.

Reply:
We discussed about a possible more generic and structural fix that might come in a future iteration.

For the time being this already enables some use-cases and improve the code-base, I'm happy to go with it!